docker/golang-cross
# Change Dockerfile to a multi-stage build

This can help reducing the final image size, but also allows for overriding parameters through build-args, without having to update the Dockerfile itself (which can be useful for testing).

In addition, some steps can now be cached, which makes local testing a bit faster.

### To test this PR

Check out the PR and build the image

```bash
docker build -t dockercore/golang-cross:1.12.9 .
```

Or if you're lazy and don't want to checkout locally;

```
docker build -t dockercore/golang-cross:1.12.9 https://github.com/thaJeztah/golang-cross.git#multistage_dockerfile
```

Then, inside your fork of the docker/cli, run:

```bash
make -f docker.Makefile clean
make -f docker.Makefile cross
```

Image size is slightly smaller as before (but could be tweaked if we split runtime and build-time dependencies)


```
# Before:
dockercore/golang-cross   1.12.9              518d5e4626aa        3 weeks ago         2.44GB

# After:
dockercore/golang-cross   1.12.9              cb8efaf3e4a9        8 minutes ago      2.4GB
```

### Before

```bash
docker pull dockercore/golang-cross:1.12.9

cd ~/go/src/github.com/docker/cli
make -f docker.Makefile clean 
make -f docker.Makefile cross

```bash
docker run --rm -e VERSION=19.09.0-dev -e GITCOMMIT -e PLATFORM -e TESTFLAGS -e TESTDIRS -e GOOS -e GOARCH -e GOARM -v "/Users/sebastiaan/go/src/github.com/docker/cli":/go/src/github.com/docker/cli -v "docker-cli-dev-cache:/root/.cache/go-build"  docker-cli-cross make cross
./scripts/build/cross
Building binaries for all platforms
Building statically linked build/docker-linux-amd64
Building statically linked build/docker-linux-arm
Generating windows resources
Building build/docker-windows-amd64
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:11: digit exceeds base
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:12: digit exceeds base
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:11: digit exceeds base
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:12: digit exceeds base
Building build/docker-darwin-amd64
Building statically linked build/docker-linux-ppc64le
Building statically linked build/docker-linux-s390x
```


### After (with this PR)

```bash
docker build -t dockercore/golang-cross:1.12.9 https://github.com/thaJeztah/golang-cross.git#multistage_dockerfile

cd ~/go/src/github.com/docker/cli
make -f docker.Makefile clean 
make -f docker.Makefile cross

docker run --rm -e VERSION=19.09.0-dev -e GITCOMMIT -e PLATFORM -e TESTFLAGS -e TESTDIRS -e GOOS -e GOARCH -e GOARM -v "/Users/sebastiaan/go/src/github.com/docker/cli":/go/src/github.com/docker/cli -v "docker-cli-dev-cache:/root/.cache/go-build"  docker-cli-cross make cross
./scripts/build/cross
Building binaries for all platforms
Building statically linked build/docker-linux-amd64
Building statically linked build/docker-linux-arm
Generating windows resources
Building build/docker-windows-amd64
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:11: digit exceeds base
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:12: digit exceeds base
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:11: digit exceeds base
/go/src/github.com/docker/cli/scripts/gen/../winresources/common.rc:12: digit exceeds base
Building build/docker-darwin-amd64
Building statically linked build/docker-linux-s390x
Building statically linked build/docker-linux-ppc64le
```
